src/billing_handler.rs

```diff
@@ -0,0 +1,83 @@
+use crate::model::AppState;
+
+use actix_web::web::{Data, Json};
+use actix_web::{get, post, HttpResponse, Responder};
+use anyhow::Context;
+use k256::elliptic_curve::generic_array::sequence::Lengthen;
+use serde_json::json;
+use tiny_keccak::{Hasher, Keccak};
+
+#[derive(Debug, serde::Deserialize)]
+pub struct SigningData {
+    nonce: String,
+    txhashes: Vec<String>,
+}
+
+#[get("/billing/inspect")]
+pub async fn get_bill(appstate: Data<AppState>) -> impl Responder {
+    HttpResponse::Ok().json(json!({
+        "bill": appstate.execution_costs.lock().unwrap().clone(),
+    }))
+}
+
+#[post("/billing/export")]
+pub async fn sign_data(appstate: Data<AppState>, data: Json<SigningData>) -> impl Responder {
+    let signing_data = data.into_inner();
+    if signing_data.nonce.is_empty() {
+        return HttpResponse::BadRequest().body("Nonce must not be empty");
+    }
+
+    if signing_data.txhashes.is_empty() {
+        return HttpResponse::BadRequest().body("List of tx hashes must not be empty");
+    }
+
+    let mut bytes32_nonce = [0u8; 32];
+    if let Err(err) = hex::decode_to_slice(&signing_data.nonce, &mut bytes32_nonce) {
+        return HttpResponse::BadRequest()
+            .body(format!("Error decoding nonce into 32 bytes: {}", err));
+    }
+
+    let mut signed_data: Vec<u8> = bytes32_nonce.to_vec();
+    for txhash in signing_data.txhashes {
+        if let Some(cost) = appstate.execution_costs.lock().unwrap().remove(&txhash) {
+            let mut bytes32_txhash = [0u8; 32];
+            if let Err(err) = hex::decode_to_slice(&txhash[2..], &mut bytes32_txhash) {
+                return HttpResponse::InternalServerError().body(format!(
```

Comment:
same thing applies here, do not exit aggressively

Reply:
Yes sure. Should I exclude the tx hash from the final signed data in this case?

Reply:
yes